File: windows/Hinge.Core/SyncModel.cs
using System.Text.Json.Serialization;

namespace Hinge.Core;

public class SyncFileEntry
{
    [JsonPropertyName("relativePath")]
    public string RelativePath { get; set; } = string.Empty;

    [JsonPropertyName("size")]
    public long Size { get; set; }

    [JsonPropertyName("modifiedTime")]
    public long ModifiedTime { get; set; }

    [JsonPropertyName("sha256")]
    public string Sha256 { get; set; } = string.Empty;

    [JsonPropertyName("platformFileIdentity")]
    public string? PlatformFileIdentity { get; set; }
}

public class SyncManifestMessage
{
    [JsonPropertyName("folderId")]
    public string FolderId { get; set; } = string.Empty;

    [JsonPropertyName("timestamp")]
    public long Timestamp { get; set; } = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

    [JsonPropertyName("entries")]
    public List<SyncFileEntry> Entries { get; set; } = new();
}

public class SyncPullRequestMessage
{
    [JsonPropertyName("folderId")]
    public string FolderId { get; set; } = string.Empty;

    [JsonPropertyName("relativePaths")]
    public List<string> RelativePaths { get; set; } = new();
}

public class SyncDifference
{
    public List<SyncFileEntry> NeedPull { get; set; } = new();
    public List<SyncFileEntry> NeedPush { get; set; } = new();
    public List<SyncFileEntry> UpToDate { get; set; } = new();
}

File: android/lib/core/sync_model.dart
/// Sync File Entry Model (Phase 5)
class SyncFileEntry {
  final String relativePath;
  final int size;
  final int modifiedTime;
  final String sha256;
  final String? platformFileIdentity;

  SyncFileEntry({
    required this.relativePath,
    required this.size,
    required this.modifiedTime,
    required this.sha256,
    this.platformFileIdentity,
  });

  Map<String, dynamic> toJson() => {
    'relativePath': relativePath,
    'size': size,
    'modifiedTime': modifiedTime,
    'sha256': sha256,
    if (platformFileIdentity != null)
      'platformFileIdentity': platformFileIdentity,
  };

  factory SyncFileEntry.fromJson(Map<String, dynamic> json) {
    return SyncFileEntry(
      relativePath: json['relativePath'] as String? ?? '',
      size: (json['size'] as num?)?.toInt() ?? 0,
      modifiedTime: (json['modifiedTime'] as num?)?.toInt() ?? 0,
      sha256: json['sha256'] as String? ?? '',
      platformFileIdentity: json['platformFileIdentity'] as String?,
    );
  }
}

/// Sync Manifest Message Model (Phase 5)
class SyncManifestMessage {
  final String folderId;
  final int timestamp;
  final List<SyncFileEntry> entries;

  SyncManifestMessage({
    required this.folderId,
    required this.timestamp,
    required this.entries,
  });

  Map<String, dynamic> toJson() => {
    'folderId': folderId,
    'timestamp': timestamp,
    'entries': entries.map((e) => e.toJson()).toList(),
  };

  factory SyncManifestMessage.fromJson(Map<String, dynamic> json) {
    final rawEntries = json['entries'] as List<dynamic>? ?? [];
    return SyncManifestMessage(
      folderId: json['folderId'] as String? ?? '',
      timestamp: (json['timestamp'] as num?)?.toInt() ?? 0,
      entries: rawEntries
          .map((e) => SyncFileEntry.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}

/// Sync Pull Request Message Model (Phase 5)
class SyncPullRequestMessage {
  final String folderId;
  final List<String> relativePaths;

  SyncPullRequestMessage({required this.folderId, required this.relativePaths});

  Map<String, dynamic> toJson() => {
    'folderId': folderId,
    'relativePaths': relativePaths,
  };

  factory SyncPullRequestMessage.fromJson(Map<String, dynamic> json) {
    final rawPaths = json['relativePaths'] as List<dynamic>? ?? [];
    return SyncPullRequestMessage(
      folderId: json['folderId'] as String? ?? '',
      relativePaths: rawPaths.map((e) => e.toString()).toList(),
    );
  }
}

/// Computed difference between local and remote manifests
class SyncDifference {
  final List<SyncFileEntry> needPull;
  final List<SyncFileEntry> needPush;
  final List<SyncFileEntry> upToDate;

  SyncDifference({
    required this.needPull,
    required this.needPush,
    required this.upToDate,
  });
}
